bundle/src/main/java/com/adobe/acs/commons/remoteassets/impl/RemoteAssetsBinarySyncImpl.java

```diff
@@ -0,0 +1,209 @@
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.HttpURLConnection;
+import java.net.URI;
+import java.net.URL;
+import java.util.GregorianCalendar;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+
+/**
+ * Service to sync a remote asset's binaries a from remote server. Implements {@link RemoteAssetsBinarySync}.
+ */
+@Component(
+    label = "ACS AEM Commons - Remote Assets - Binary Sync Service",
+    description = "Pulls the binaries for a remote asset in order to make it a true local asset."
+)
+@Service
+public class RemoteAssetsBinarySyncImpl implements RemoteAssetsBinarySync {
+
+    private static final Logger LOG = LoggerFactory.getLogger(RemoteAssetsBinarySyncImpl.class);
+
+    @Reference
+    private ResourceResolverFactory resourceResolverFactory;
+
+    @Reference
+    private RemoteAssetsConfig remoteAssetsConfig;
+
+    private ResourceResolver resourceResolver;
+    private Session session;
+
+    /**
+     * Method to run on activation.
+     * @throws RepositoryException exception
+     */
+    @Activate
+    protected void activate() throws RepositoryException {
+        this.resourceResolver = RemoteAssets.logIn(this.resourceResolverFactory);
+        this.session = this.resourceResolver.adaptTo(Session.class);
+        if (StringUtils.isNotBlank(this.remoteAssetsConfig.getEventUserData())) {
+            this.session.getWorkspace().getObservationManager().setUserData(this.remoteAssetsConfig.getEventUserData());
+        }
+    }
+
+    /**
+     * Method to run on deactivation.
+     */
+    @Deactivate
+    protected void deactivate() {
+        if (this.session != null) {
+            try {
+                this.session.logout();
+            } catch (Exception e) {
+                LOG.warn("Failed session.logout()", e);
+            }
+        }
+        if (this.resourceResolver != null) {
+            try {
+                this.resourceResolver.close();
+            } catch (Exception e) {
+                LOG.warn("Failed resourceResolver.close()", e);
+            }
+        }
+    }
+
+    /**
+     * @see RemoteAssetsBinarySync#syncAsset(Resource)
+     * @param resource Resource
+     * @return Resource
+     */
+    @Override
+    public Resource syncAsset(Resource resource) {
+        try {
+            this.session.refresh(true);
+            Resource localRes = this.resourceResolver.getResource(resource.getPath());
+            Node node = localRes.adaptTo(Node.class);
+
+            Asset asset = localRes.getParent().adaptTo(Asset.class);
+            URI pathUi = new URI(null, null, asset.getPath(), null);
+            String baseUrl = this.remoteAssetsConfig.getServer().concat(pathUi.toString()).concat("/_jcr_content/renditions/");
+
+            Iterator<? extends Rendition> renditions = asset.listRenditions();
+            while (renditions.hasNext()) {
+                Rendition assetRendition = renditions.next();
```

Comment:
Im thinking these HTTP calls could get passed off to fast action manager to queue up and manage the work; it wont run away from you like if you were spawning threads, and it doesnt have the overhead of sling jobs.

Making the renditions sync synchronous on the request feels like trouble waiting to happen.

Reply:
I think this is a good consideration for a future iteration, where we monitor how many of these HTTP calls have queued up and are waiting, delegating to asynchronous sync when the queue gets a certain length.

Synchronous pull is what makes this function truly "magical".  Without it, the first time you access an asset you'd see the temporary binary.  The user would need to refresh the page to get the real thing, stealing much of the appeal of this feature.

If we add (again in a later enhancement) a feature to make requests async after X number are currently active, then a user could configure this to 0 if they truly want ALL syncs to be async.